<a href="https://colab.research.google.com/github/Sadanan643021272-0/BSC_DPDM23/blob/main/Copy_of_Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [9]:
import pandas as pd
import os

In [10]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/My Drive/bsc_dpdm23_data'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [11]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/BSC_DPDM23/reduced_marketbasket.csv'),encoding='latin-1')

In [12]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [13]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [14]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [15]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [16]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=c2569dc0678e0dd7cb87237c9bf4cd37128feada09d29af2da5d0f1ac95a8aca
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [17]:
from apyori import apriori

In [29]:
list(apriori(transactions,min_support = 0.01))

[RelationRecord(items=frozenset({'15036'}), support=0.020193050193050194, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'15036'}), confidence=0.020193050193050194, lift=1.0)]),
 RelationRecord(items=frozenset({'15056BL'}), support=0.012586872586872588, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'15056BL'}), confidence=0.012586872586872588, lift=1.0)]),
 RelationRecord(items=frozenset({'15056N'}), support=0.017876447876447876, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'15056N'}), confidence=0.017876447876447876, lift=1.0)]),
 RelationRecord(items=frozenset({'16237'}), support=0.011235521235521236, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'16237'}), confidence=0.011235521235521236, lift=1.0)]),
 RelationRecord(items=frozenset({'20675'}), support=0.012509652509652509, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_a

In [18]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [19]:
len(frequentItemsets)

4

In [20]:
frequentItemsets[0][0]

frozenset({'22423'})

In [21]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [22]:
frequentItemsets[0][0].__len__()

1

In [23]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [24]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [25]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'575138',
 '576181',
 '546688',
 '537738',
 '564974',
 '544779',
 '541999',
 '561702',
 '566720',
 '557735',
 '563023',
 '548663',
 '539465',
 '567157',
 '565150',
 '559112',
 '545601',
 '550437',
 '560036',
 '547096',
 '581122',
 '537435',
 '566790',
 '567184',
 '565919',
 '552843',
 '564306',
 'C545444',
 '555522',
 '561070',
 '545701',
 '568567',
 'C551511',
 '546888',
 '569545',
 '541110',
 '542881',
 '537042',
 '564857',
 '568792',
 '581170',
 '541421',
 '554337',
 'C550664',
 '562711',
 '539434',
 '541949',
 '559693',
 '549746',
 '567639',
 '578273',
 '566193',
 '561631',
 '556206',
 '570086',
 '541709',
 '560790',
 '567670',
 '573330',
 '542999',
 '559557',
 'C564964',
 '540352',
 '539436',
 '551352',
 '540371',
 '554028',
 '550205',
 '569570',
 '538148',
 '578943',
 'C555409',
 '575703',
 '557350',
 '559357',
 '566326',
 '556016',
 '554921',
 '550468',
 '552315',
 'C556793',
 '556503',
 'C567955',
 '574368',
 '547400',
 '564541',
 '543116',
 '536804',
 '572751',
 '571283',
 '5

In [26]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

##แปลงData ให้อยู่ในรูปแบบ Transaction data ด้วย groupby (list)

In [27]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [28]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

##quiz 22386 และ 85099B คือสินค้าอะไร

In [33]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B')]['InvoiceNO'])

KeyError: ignored

## HW1 หา frequency Item Set และ Association  Rules
--ของช่วงเวลาคริสมาร์ส&ปีใหม่ (1-31ธันวาคม2566)

--ของประเทศในยุโรป

--ของช่วงเวลา ตรุษจีน ในทวีปเอเชีย (15 มค ถึง 15 กพ)

--ของวันวาเลนไทน์

--ของช่วงเวลาคริสมาร์ส&ปีใหม่ ในทวีปยุโรป (1 ธค ถึง 31 ธค)

--โดยจัดกลุ่มตามเดือน ประเทศ
